File: tests/test_heuristic.py
"""Test heuristics for the prize collecting TSP"""

import networkx as nx
import pytest
from tspwplib import (
    asymmetric_from_undirected,
    biggest_vertex_id_from_graph,
    edge_list_from_walk,
    is_simple_cycle,
    split_head,
    total_cost,
    total_cost_networkx,
    total_prize,
    total_prize_of_tour,
    VertexFunctionName,
)
from pctsp.algorithms import (
    collapse,
    extension_until_prize_feasible,
    find_cycle_from_bfs,
    path_extension_collapse,
    path_extension_until_prize_feasible,
    random_tour_complete_graph,
    suurballes_heuristic,
    suurballes_tour_initialization,
    tour_from_vertex_disjoint_paths,
)
from pctsp.algorithms.extension_collapse import (
    extension_unitary_gain,
    extension_unitary_loss,
)
from pctsp.preprocessing import (
    undirected_vertex_disjoint_paths_map,
    vertex_disjoint_cost_map,
)
from pctsp.suurballe import suurballe_shortest_vertex_disjoint_paths


def test_collapse(suurballes_undirected_graph, root):
    """Test if a tour is collapsed"""
    tour = [0, 1, 4, 6, 7, 2, 0]
    quota = 5
    new_tour = collapse(suurballes_undirected_graph, tour, quota, root)
    assert total_cost_networkx(suurballes_undirected_graph, new_tour) == 16
    assert (
        total_prize_of_tour(
            nx.get_node_attributes(suurballes_undirected_graph, "prize"), new_tour
        )
        >= quota
    )
    assert new_tour[0] == new_tour[len(new_tour) - 1] == root


def test_extension_unitary_gain(suurballes_undirected_graph):
    """Test if a tour is extended"""
    tour = [0, 1, 3, 6, 7, 2, 0]
    extended_tour = extension_unitary_gain(suurballes_undirected_graph, tour)
    assert 4 not in extended_tour
    assert 5 in extended_tour
    assert len(extended_tour) == len(tour) + 1


def test_extension_unitary_loss(suurballes_undirected_graph, root):
    """Test if a tour is extended"""
    tour = [0, 1, 3, 6, 7, 2, 0]
    extended_tour = extension_unitary_loss(suurballes_undirected_graph, tour, root)
    assert 4 not in extended_tour
    assert 5 in extended_tour
    assert len(extended_tour) == len(tour) + 1


def test_extension_tsplib(tspwplib_graph, root):
    """Test if a tour is extended on the tsplib dataset"""
    n = tspwplib_graph.number_of_nodes()
    tour = [0, 1, 2, n - 1, n - 2, 0]
    extended_tour = extension_unitary_loss(tspwplib_graph, tour, root)
    prize_map = nx.get_node_attributes(tspwplib_graph, VertexFunctionName.prize.value)
    assert total_prize_of_tour(prize_map, extended_tour) > total_prize_of_tour(
        prize_map, tour
    )
    assert root in extended_tour
    assert is_simple_cycle(tspwplib_graph, extended_tour)
    for u in tspwplib_graph:
        assert extended_tour.count(u) < 2 or u == root


def test_extension_until_prize_feasible(tspwplib_graph, root):
    """Run extension with unitary gain until the tour is prize feasible"""
    n = tspwplib_graph.number_of_nodes()
    prize_map = nx.get_node_attributes(tspwplib_graph, VertexFunctionName.prize.value)

    # set quota to be the total prize of the graph, i.e. extended tour is Hamiltonian
    quota = total_prize(prize_map, tspwplib_graph.nodes())
    tour = [0, 1, 2, n - 1, n - 2, 0]
    extended_tour = extension_until_prize_feasible(
        tspwplib_graph,
        tour,
        quota,
    )
    assert total_prize_of_tour(prize_map, extended_tour) >= quota
    assert root in extended_tour
    assert is_simple_cycle(tspwplib_graph, extended_tour)
    for u in tspwplib_graph:
        assert extended_tour.count(u) < 2 or u == root


@pytest.mark.parametrize("step_size,path_depth_limit", [(1, 2), (1, 4)])
def test_path_extension_until_feasible(
    tspwplib_graph, root, step_size, path_depth_limit
):
    """Test obtaining a feasible tour via extension"""
    quota = 20
    n = tspwplib_graph.number_of_nodes()
    tour = [0, 1, 2, n - 1, n - 2, 0]
    extended_tour = path_extension_until_prize_feasible(
        tspwplib_graph,
        tour,
        root,
        quota,
        step_size=step_size,
        path_depth_limit=path_depth_limit,
    )
    prize_map = nx.get_node_attributes(tspwplib_graph, VertexFunctionName.prize.value)
    assert total_prize_of_tour(prize_map, extended_tour) >= quota
    assert root in extended_tour
    assert is_simple_cycle(tspwplib_graph, extended_tour)
    for u in tspwplib_graph:
        assert extended_tour.count(u) < 2 or u == root


def test_path_extension_collapse(suurballes_undirected_graph, root):
    """Test if a tour is extended and collapsed using paths"""
    tour = [0, 1, 3, 6, 7, 2, 0]
    quota = 7
    new_tour = path_extension_collapse(
        suurballes_undirected_graph,
        tour,
        root,
        quota,
        collapse_shortest_paths=True,
        path_depth_limit=suurballes_undirected_graph.number_of_nodes(),
        step_size=10,
    )
    assert (
        total_prize_of_tour(
            nx.get_node_attributes(suurballes_undirected_graph, "prize"), new_tour
        )
        >= quota
    )


@pytest.mark.parametrize("step_size", [1, 2, 3])
def test_pec_tspwplib(tspwplib_graph, root, step_size):
    """Test obtaining a feasible tour via extension"""
    quota = 20
    n = tspwplib_graph.number_of_nodes()
    tour = [0, 1, 2, n - 1, n - 2, 0]
    extended_tour = path_extension_collapse(
        tspwplib_graph,
        tour,
        root,
        quota,
        collapse_shortest_paths=True,
        path_depth_limit=tspwplib_graph.number_of_nodes(),
        step_size=step_size,
    )
    prize_map = nx.get_node_attributes(tspwplib_graph, VertexFunctionName.prize.value)
    assert total_prize_of_tour(prize_map, extended_tour) >= quota
    assert root in extended_tour
    assert is_simple_cycle(tspwplib_graph, extended_tour)
    for u in tspwplib_graph:
        assert extended_tour.count(u) < 2 or u == root


def test_random_tour_complete_graph(tspwplib_graph, root):
    """Test random tours on complete graphs"""
    prize_dict = nx.get_node_attributes(tspwplib_graph, VertexFunctionName.prize.value)
    prize = total_prize(prize_dict, tspwplib_graph.nodes())
    quota = int(float(prize) * 0.5)
    tour = random_tour_complete_graph(tspwplib_graph, root, quota)
    assert total_prize_of_tour(prize_dict, tour) >= quota


def test_suurballes_heuristic(suurballes_undirected_graph, root):
    """Test Suurballe's heuristic"""
    # setup params
    quota = 5
    prize_map = dict(suurballes_undirected_graph.nodes(data="prize"))

    # convert to asymmetric graph and run Suurballe's
    biggest_vertex = biggest_vertex_id_from_graph(suurballes_undirected_graph)
    asymmetric_graph = asymmetric_from_undirected(suurballes_undirected_graph)
    assert (
        asymmetric_graph.number_of_edges()
        == 2 * suurballes_undirected_graph.number_of_edges()
        + suurballes_undirected_graph.number_of_nodes()
    )
    tree = suurballe_shortest_vertex_disjoint_paths(
        asymmetric_graph, split_head(biggest_vertex, root), weight="cost"
    )
    cost_map = vertex_disjoint_cost_map(tree, biggest_vertex)
    vertex_disjoint_paths_map = undirected_vertex_disjoint_paths_map(
        tree, biggest_vertex
    )

    # run the heuristic
    tour = suurballes_heuristic(prize_map, quota, cost_map, vertex_disjoint_paths_map)

    assert tour in ([0, 1, 5, 7, 2, 0], [0, 1, 5, 2, 0])
    assert (
        total_cost(
            nx.get_edge_attributes(suurballes_undirected_graph, "cost"),
            edge_list_from_walk(tour),
        )
        == 16
    )
    assert total_prize_of_tour(prize_map, tour) >= quota


@pytest.mark.parametrize("alpha", ([25, 50, 75]))
def test_suurballes_tour_initialization(sparse_tspwplib_graph, root, alpha) -> None:
    """Test finding an initial tour with Suurballe's algorithm"""
    prize_map = dict(sparse_tspwplib_graph.nodes(data="prize"))
    quota = int(sum(prize_map.values()) * alpha * 0.01)
    print("Quota:", quota)
    for u in sparse_tspwplib_graph:
        if sparse_tspwplib_graph.has_edge(u, u):
            sparse_tspwplib_graph.remove_edge(u, u)

    # convert to asymmetric graph and run Suurballe's
    biggest_vertex = biggest_vertex_id_from_graph(sparse_tspwplib_graph)
    asymmetric_graph = asymmetric_from_undirected(sparse_tspwplib_graph)
    assert nx.number_of_selfloops(asymmetric_graph) == 0
    assert (
        asymmetric_graph.number_of_edges()
        == 2 * sparse_tspwplib_graph.number_of_edges()
        + sparse_tspwplib_graph.number_of_nodes()
    )
    tree = suurballe_shortest_vertex_disjoint_paths(
        asymmetric_graph, split_head(biggest_vertex, root), weight="cost"
    )
    cost_map = vertex_disjoint_cost_map(tree, biggest_vertex)
    vertex_disjoint_paths_map = undirected_vertex_disjoint_paths_map(
        tree, biggest_vertex
    )
    tour = suurballes_tour_initialization(
        prize_map, quota, cost_map, vertex_disjoint_paths_map
    )
    prize_of_tour = total_prize_of_tour(prize_map, tour)
    if prize_of_tour < quota:
        for disjoint_paths in vertex_disjoint_paths_map.values():
            tour_from_paths = tour_from_vertex_disjoint_paths(disjoint_paths)
            assert total_prize_of_tour(prize_map, tour_from_paths) <= prize_of_tour
    else:
        cost_of_tour = total_cost(
            nx.get_edge_attributes(sparse_tspwplib_graph, "cost"),
            edge_list_from_walk(tour),
        )
        for vertex, disjoint_paths in vertex_disjoint_paths_map.items():
            tour_from_paths = tour_from_vertex_disjoint_paths(disjoint_paths)
            if total_prize_of_tour(prize_map, tour_from_paths) < quota:
                assert cost_map[vertex] >= cost_of_tour


def test_find_cycle_from_bfs(suurballes_undirected_graph, root):
    """Test finding a simple cycle including the root"""
    cycle = find_cycle_from_bfs(suurballes_undirected_graph, root)
    assert is_simple_cycle(suurballes_undirected_graph, cycle)
    assert root in cycle
    assert cycle[0] == root
    assert cycle[len(cycle) - 1] == root


def test_find_cycle_from_bfs_tsplib(tspwplib_graph, root):
    """Test finding simple cycle on tsplib graph"""
    cycle = find_cycle_from_bfs(tspwplib_graph, root)
    assert is_simple_cycle(tspwplib_graph, cycle)
    assert root in cycle
    assert cycle[0] == root
    assert cycle[len(cycle) - 1] == root

File: tests/test_heuristic.cpp
#include <algorithm>
#include <boost/graph/adjacency_list.hpp>
#include <boost/graph/dijkstra_shortest_paths_no_color_map.hpp>
#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include <iostream>
#include <typeinfo>

#include "fixtures.hh"
#include "pctsp/heuristic.hh"

using namespace std;
using namespace boost;

int expected_num_edges_in_complete_graph(int n_vertices) {
    return (n_vertices * (n_vertices - 1)) / 2;
}

typedef GraphFixture CompleteGraphParameterizedFixture;
typedef GraphFixture SuurballeGraphFixture;
typedef GraphFixture HeuristicFixture;

TEST(TestExtensionCollapse, testUnitaryGain) {
    EXPECT_EQ(unitaryGain(10, 2, 2, 2), 5);
    EXPECT_EQ(unitaryGain(2, 0, 1, 3), 0.5);
    // triangle inequality does not hold
    EXPECT_EQ(unitaryGain(10, 15, 5, 5), -2);
}

TEST(TestExtensionCollapse, testCalculateAverageGain) {
    typedef std::map<int, ExtensionVertex> UnitaryGainMap;
    UnitaryGainMap gain_map;

    std::unordered_set<int> vertices_in_tour = { 0, 1, 2, 3, 0 };
    EXPECT_EQ(vertices_in_tour.size(), 4);

    // vertices not in the tour
    int a = 4;
    int b = 5;
    int c = 6;

    // initialise unitary gain and gain map
    ExtensionVertex gain_a = ExtensionVertex();
    ExtensionVertex gain_b = ExtensionVertex();
    ExtensionVertex gain_c = ExtensionVertex();

    gain_a.value = 1;
    gain_a.index = 1;
    gain_b.value = 2;
    gain_b.index = 2;
    gain_c.value = 6;
    gain_c.index = 3;

    gain_map[a] = gain_a;
    gain_map[b] = gain_b;
    gain_map[c] = gain_c;

    // expect average gain to be 4
    EXPECT_EQ(calculateAverageGain(vertices_in_tour, gain_map), 3);

    // now add a to the tour and re-calcuate avg gain (should not include gain
    // of a)
    vertices_in_tour.insert(a);
    EXPECT_EQ(calculateAverageGain(vertices_in_tour, gain_map), 4);
}

TEST_P(CompleteGraphParameterizedFixture, testUnitaryGainOfVertex) {
    typedef typename PCTSPgraph::vertex_descriptor Vertex;
    PCTSPgraph g = getGraph();
    Vertex v0 = boost::vertex(0, g);
    Vertex v1 = boost::vertex(1, g);
    Vertex v2 = boost::vertex(2, g);
    // std::list<Vertex> tour = {v0, v1, v2, v0};
    std::list<PCTSPvertex> tour = { 0, 1, 2, 0 };
    auto prize_map = getPrizeMap(g);
    auto cost_map = getCostMap(g);
    // Vertex missing_vertex = boost::vertex(3, g);
    int missing_vertex = 3;
    ExtensionVertex gain =
        unitaryGainOfVertex(g, tour, cost_map, prize_map, missing_vertex);
    if (num_vertices(g) == 4) {
        EXPECT_FLOAT_EQ(0.5, gain.value);
    }
    else if (num_vertices(g) == 5) {
        EXPECT_FLOAT_EQ(3.0 / 7.0, gain.value);
    }
}

TEST_P(HeuristicFixture, testExtensionUnitaryGain) {
    PCTSPgraph graph = getGraph();
    auto prize_map = getPrizeMap(graph);
    auto cost_map = getCostMap(graph);
    auto quota = getQuota();
    auto root_vertex = getRootVertex();
    std::list<PCTSPvertex> tour =getPrizeFeasibleTour();
    auto prize_before = totalPrizeOfTour(prize_map, tour);
    extensionUnitaryGain(graph, tour, cost_map, prize_map);
    auto total_prize = totalPrizeOfTour(prize_map, tour);
    EXPECT_GE(total_prize, quota);
    switch(GetParam()) {
        case GraphType::COMPLETE4:
        case GraphType::GRID8:
            EXPECT_EQ(total_prize, prize_before); break;
        case GraphType::COMPLETE5:
        case GraphType::SUURBALLE:
            EXPECT_GT(total_prize, prize_before); break;
        default: break;
    }
}

TEST_P(SuurballeGraphFixture, testExtensionUnitaryGain) {
    PCTSPgraph graph = getGraph();
    std::list<PCTSPvertex> tour = { 0, 1, 3, 6, 7, 2, 0 };
    int tour_size_before_extend = tour.size();

    // get the edge property map from bundled internal property
    auto prize_map = getPrizeMap(graph);
    auto cost_map = getCostMap(graph);
    extensionUnitaryGain(graph, tour, cost_map, prize_map);

    // we expect that vertex e (index 5) has been added, but not vertex d
    EXPECT_EQ(tour.size(), tour_size_before_extend + 1);
    int d = 4;
    int e = 5;

    // check vertex d not in tour
    auto d_iterator = std::find(tour.begin(), tour.end(), d);
    EXPECT_TRUE(d_iterator == tour.end());

    // check e is added to the tour in the correct index
    auto e_iterator = std::find(tour.begin(), tour.end(), e);
    int e_index = std::distance(tour.begin(), e_iterator);
    EXPECT_FALSE(e_iterator == tour.end());
    EXPECT_EQ(e_index, 5);
}

TEST_P(HeuristicFixture, testExtensionUntilPrizeFeasible) {
    PCTSPgraph graph = getGraph();
    auto prize_map = getPrizeMap(graph);
    auto cost_map = getCostMap(graph);
    auto quota = getQuota();
    auto root_vertex = getRootVertex();
    std::list<PCTSPvertex> tour = getSmallTour();

    // run the algorithm
    extensionUntilPrizeFeasible(graph, tour, cost_map, prize_map, quota);

    // we expect the total prize of the tour to be at least the quota
    auto total_prize = totalPrizeOfTour(prize_map, tour);
    switch (GetParam()) {
        case GraphType::GRID8: EXPECT_LT(total_prize, quota); break;
        default: EXPECT_GE(total_prize, quota);
    }
}

TEST_P(HeuristicFixture, testCollapse) {
    // get graphs with property maps
    PCTSPgraph graph = getGraph();
    auto prize_map = getPrizeMap(graph);
    auto cost_map = getCostMap(graph);
    auto quota = getQuota();
    auto root_vertex = getRootVertex();
    std::list<PCTSPvertex> tour;
    std::list<PCTSPvertex> expected_collapse;

    switch (GetParam()) {
        case GraphType::SUURBALLE: {
            // this tour can be collapsed to {0, 1, 5, 2, 0}
            quota = 4;
            tour = { 0, 1, 4, 6, 7, 2, 0 };
            expected_collapse = {0, 1, 5, 2, 0};
            break;
        }
        case GraphType::GRID8: {
            tour = { 0, 1, 4, 6, 7, 5, 3, 2, 0};
            expected_collapse = { 0, 1, 4, 5, 3, 2, 0};
            break;
        }
        case GraphType::COMPLETE4: {
            tour = { 0, 1, 2, 3, 0 };
            expected_collapse = {0, 1, 3, 0};
            break;
        }
        case GraphType::COMPLETE5: {
            tour = { 0, 1, 2, 3, 4, 0 };
            expected_collapse = {0, 1, 2, 3, 0};
            break;
        }
        default: {
            tour = {};
            expected_collapse = {};
            break;
        }
    }

    // run the collapse heuristic
    std::list<PCTSPvertex> new_tour = collapse(graph, tour, cost_map, prize_map, quota, root_vertex);

    auto expected_cost = totalCost(graph, expected_collapse, cost_map);
    auto actual_cost = totalCost(graph, new_tour, cost_map);
    auto expected_prize = totalPrizeOfTour(prize_map, expected_collapse);
    auto actual_prize = totalPrizeOfTour(prize_map, new_tour);
    EXPECT_EQ(actual_cost, expected_cost);
    EXPECT_GE(actual_prize, quota);
    EXPECT_EQ(actual_prize, expected_prize);
    EXPECT_EQ(actual_cost, expected_cost);

    // ensure the first vertex of the tour is the root vertex
    auto first_root_it =
        std::find(new_tour.begin(), new_tour.end(), root_vertex);
    int root_index = std::distance(new_tour.begin(), first_root_it);
    EXPECT_FALSE(first_root_it == new_tour.end());
    EXPECT_EQ(root_index, 0);

    // ensure the last vertex of the tour is the root vertex
    ++first_root_it;
    auto second_root_it = std::find(first_root_it, tour.end(), root_vertex);
    int second_root_index = std::distance(new_tour.begin(), second_root_it);
    EXPECT_FALSE(second_root_it == new_tour.end());
    EXPECT_EQ(second_root_index, new_tour.size() - 1);
}


TEST_P(HeuristicFixture, testCollapseShortestPath) {
    // get graphs with property maps
    PCTSPgraph graph = getGraph();
    auto prize_map = getPrizeMap(graph);
    auto cost_map = getCostMap(graph);
    auto quota = getQuota();
    auto root_vertex = getRootVertex();
    std::list<PCTSPvertex> tour;
    std::list<PCTSPvertex> expected_collapse;

    switch (GetParam()) {
        case GraphType::SUURBALLE: {
            // this tour can be collapsed to {0, 1, 5, 2, 0}
            quota = 4;
            tour = { 0, 1, 4, 6, 7, 2, 0 };
            expected_collapse = {0, 1, 5, 2, 0};
            break;
        }
        case GraphType::GRID8: {
            tour = { 0, 1, 4, 6, 7, 5, 3, 2, 0};
            expected_collapse = { 0, 1, 4, 5, 3, 2, 0};
            break;
        }
        case GraphType::COMPLETE4: {
            tour = { 0, 1, 2, 3, 0 };
            expected_collapse = {0, 1, 3, 0};
            break;
        }
        case GraphType::COMPLETE5: {
            tour = { 0, 1, 2, 3, 4, 0 };
            expected_collapse = {0, 1, 2, 3, 0};
            break;
        }
        default: {
            tour = {};
            expected_collapse = {};
            break;
        }
    }

    // run the collapse heuristic
    std::list<PCTSPvertex> new_tour = collapse(graph, tour, cost_map, prize_map, quota, root_vertex, true);

    auto expected_cost = totalCost(graph, expected_collapse, cost_map);
    auto actual_cost = totalCost(graph, new_tour, cost_map);
    auto expected_prize = totalPrizeOfTour(prize_map, expected_collapse);
    auto actual_prize = totalPrizeOfTour(prize_map, new_tour);
    EXPECT_EQ(actual_cost, expected_cost);
    EXPECT_GE(actual_prize, quota);
    EXPECT_EQ(actual_prize, expected_prize);
    EXPECT_EQ(actual_cost, expected_cost);

    // ensure the first vertex of the tour is the root vertex
    auto first_root_it =
        std::find(new_tour.begin(), new_tour.end(), root_vertex);
    int root_index = std::distance(new_tour.begin(), first_root_it);
    EXPECT_FALSE(first_root_it == new_tour.end());
    EXPECT_EQ(root_index, 0);

    // ensure the last vertex of the tour is the root vertex
    ++first_root_it;
    auto second_root_it = std::find(first_root_it, tour.end(), root_vertex);
    int second_root_index = std::distance(new_tour.begin(), second_root_it);
    EXPECT_FALSE(second_root_it == new_tour.end());
    EXPECT_EQ(second_root_index, new_tour.size() - 1);
}

TEST(TestExtensionCollapse, testIndexOfReverseIterator) {
    std::list<int> mylist = { 0, 1, 2, 3, 4, 0 };
    auto rit = mylist.rbegin();
    EXPECT_EQ(indexOfReverseIterator(mylist, rit), 5);
    ++rit;
    ++rit;
    EXPECT_EQ(indexOfReverseIterator(mylist, rit), 3);
    ++rit;
    ++rit;
    ++rit;
    EXPECT_EQ(indexOfReverseIterator(mylist, rit), 0);
}

TEST_P(HeuristicFixture, testNeighborIntersection) {
    PCTSPgraph graph = getGraph();
    auto u = boost::vertex(0, graph);
    auto v = boost::vertex(1, graph);
    auto intersection = neighborIntersection(graph, u, v);
    int expected_size;
    switch (GetParam()) {
        case GraphType::COMPLETE4:
            expected_size = 2;
            break;
        case GraphType::COMPLETE5: {
            expected_size = 3;
            break;
        }
        case GraphType::GRID8: {
            expected_size = 0;
            break;
        }
        case GraphType::SUURBALLE: {
            expected_size = 1;
            break;
        }
        default: {
            expected_size = 0;
            break;
        }
    }
    EXPECT_EQ(expected_size, intersection.size());
}

TEST_P(HeuristicFixture, testExtensionUnitaryLoss) {
    PCTSPgraph graph = getGraph();
    auto cost_map = getCostMap(graph);
    auto prize_map = getPrizeMap(graph);
    auto small_tour = getSmallTour();
    auto root = getRootVertex();
    if (GetParam() == GraphType::SUURBALLE) {
        small_tour = {0, 1, 3, 6, 7, 2, 0};
    }
    auto old_size = small_tour.size();
    int expected_size;
    extensionUnitaryLoss(graph, small_tour, cost_map, prize_map, root);
    switch (GetParam()) {
        case GraphType::SUURBALLE:
        case GraphType::COMPLETE5:
            expected_size = old_size + 1; break;
        default:
            expected_size = old_size;
    }
    EXPECT_EQ(small_tour.size(), expected_size);
}

TEST_P(HeuristicFixture, testExtensionStep) {
    PCTSPgraph graph = getGraph();
    auto cost_map = getCostMap(graph);
    auto prize_map = getPrizeMap(graph);
    auto small_tour = getSmallTour();
    auto root = getRootVertex();
    auto old_size = small_tour.size();
    int expected_size;
    int step_size = 2;
    int path_depth_limit = 2;
    pathExtension(graph, small_tour, cost_map, prize_map, root, step_size, path_depth_limit);
    switch (GetParam()) {
        default:
            expected_size = old_size; break;
    }
    EXPECT_EQ(small_tour.size(), expected_size);
}

TEST_P(HeuristicFixture, testPathExtensionUntilPrizeFeasible) {
    auto graph = getGraph();
    auto cost_map = getCostMap(graph);
    auto prize_map = getPrizeMap(graph);
    auto root = getRootVertex();
    auto small_tour = getSmallTour();
    auto quota = getQuota();

    int step_size = 1;
    int path_depth_limit = 2;

    pathExtensionUntilPrizeFeasible(graph, small_tour, cost_map, prize_map, root, quota, step_size, path_depth_limit);

    switch (GetParam()) {
        case GraphType::GRID8: EXPECT_LT(totalPrizeOfTour(prize_map, small_tour), quota); break;
        default: EXPECT_GE(totalPrizeOfTour(prize_map, small_tour), quota); break;
    }
}

TEST_P(HeuristicFixture, testSwapPathsInTour) {
    auto small_tour = getSmallTour();
    auto old_length = small_tour.size();
    std::list<PCTSPvertex> new_path = {5, 8, 0, 10};
    // when i < j is the easy test
    int i = 1;
    int j = 3;
    swapPathsInTour(small_tour, new_path, i, j);
    EXPECT_EQ(old_length - (j - i + 1) + new_path.size(), small_tour.size());
}


TEST_P(HeuristicFixture, testSwapPathsInTourWithRoot) {
    auto small_tour = getSmallTour();
    auto old_length = small_tour.size();
    std::list<PCTSPvertex> new_path = {5, 8, 0, 10};
    // when i > j is more difficult
    int i = 3;
    int j = 1;
    swapPathsInTour(small_tour, new_path, i, j);
    EXPECT_EQ(old_length + new_path.size() + 1 - (old_length - i + j + 1), small_tour.size());
}

TEST_P(HeuristicFixture, testExtensionPathDepth) {
    auto graph = getGraph();
    auto cost_map = getCostMap(graph);
    auto prize_map = getPrizeMap(graph);
    auto root = getRootVertex();
    auto small_tour = getSmallTour();
    auto quota = getQuota();

    int step_size = 1;
    int path_depth_limit = 3;

    pathExtensionUntilPrizeFeasible(graph, small_tour, cost_map, prize_map, root, quota, step_size, path_depth_limit);

    auto prize = totalPrizeOfTour(prize_map, small_tour);
    switch (GetParam()) {
        // case GraphType::SUURBALLE: EXPECT_LE(prize, quota); break;
        default: EXPECT_GE(prize, quota); break;
    }
}

TEST_P(HeuristicFixture, testPathExtensionCollapse) {
    auto graph = getGraph();
    auto cost_map = getCostMap(graph);
    auto prize_map = getPrizeMap(graph);
    auto root = getRootVertex();
    auto small_tour = getSmallTour();
    auto quota = getQuota();

    bool collapse_shortest_paths = false;
    int path_depth_limit = boost::num_vertices(graph);
    int step_size = 1;
    auto tour = pathExtensionCollapse(graph, small_tour, cost_map, prize_map, quota, root, collapse_shortest_paths, path_depth_limit, step_size);
    EXPECT_GE(totalPrizeOfTour(prize_map, tour), quota);

    if (totalPrizeOfTour(prize_map, small_tour) >= quota) {
        EXPECT_LE(totalCost(graph, tour, cost_map), totalCost(graph, small_tour, cost_map));
    }
}

INSTANTIATE_TEST_SUITE_P(TestExtensionCollapse, CompleteGraphParameterizedFixture,
    ::testing::Values(GraphType::COMPLETE4, GraphType::COMPLETE5)
);

INSTANTIATE_TEST_SUITE_P(TestExtensionCollapse, SuurballeGraphFixture,
    ::testing::Values(GraphType::SUURBALLE)
);

INSTANTIATE_TEST_SUITE_P(TestExtensionCollapse, HeuristicFixture,
    ::testing::Values(GraphType::SUURBALLE, GraphType::GRID8, GraphType::COMPLETE4, GraphType::COMPLETE5)
);
